### PR TITLE
🌱 Create unified yaml diff in clusterctl alpha topology plan for having a human readable output

### LINKED ISSUE #6985: clusterctl topology plan should generate diff files in yaml
**What steps did you take and what happened:**
Run clusterctl topology plan and I noticed that the .diff files represent resources as a go type instead of yaml, e.g. 

```
  &unstructured.Unstructured{
  	Object: map[string]any{
  		"apiVersion": string("controlplane.cluster.x-k8s.io/v1beta1"),
  		"kind":       string("KubeadmControlPlane"),
  		"metadata": map[string]any{
  			"annotations":       map[string]any{"cluster.x-k8s.io/cloned-from-groupkind": string("KubeadmControlPlaneTemplate.controlplane.cluster.x-k8s.io"), "cluster.x-k8s.io/cloned-from-name": string("clusterclass1-controlplane-template")},
  			"creationTimestamp": string("2022-07-26T13:34:57Z"),
  			"finalizers":        []any{string("kubeadm.controlplane.cluster.x-k8s.io")},
- 			"generation":        int64(1),
+ 			"generation":        int64(2),
```

**What did you expect to happen:**
to have a diff in yaml (easier to read)

**Environment:**

- Cluster-api version: main

/kind bug
/area topology
/area clusterctl

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

Creates a diff based on the yaml output instead of go struct output.

Prior to this change it did make use of the cmp package which did create diffs on go object level.

Old output:
```
  &unstructured.Unstructured{
        Object: map[string]any{
                "apiVersion": string("cluster.x-k8s.io/v1beta1"),
                "kind":       string("MachineDeployment"),
                "metadata": map[string]any{
                        ... // 6 identical entries
                        "namespace":       string("default"),
                        "ownerReferences": []any{map[string]any{"apiVersion": string("cluster.x-k8s.io/v1beta1"), "kind": string("Cluster"), "name": string("development-8403"), "uid": string("42383b5f-a1e4-4353-af98-54b32f0c05f3")}},
-                       "resourceVersion": string("16783"),
+                       "resourceVersion": string("2"),
                        "uid":             string("1cdbcd7e-d949-49c8-9c50-7ec5d98193a8"),
                },
                "spec": map[string]any{
                        "clusterName":             string("development-8403"),
                        "minReadySeconds":         int64(0),
                        "progressDeadlineSeconds": int64(600),
-                       "replicas":                int64(3),
+                       "replicas":                int64(4),
                        "revisionHistoryLimit":    int64(1),
                        "selector":                map[string]any{"matchLabels": map[string]any{"cluster.x-k8s.io/cluster-name": string("development-8403"), "topology.cluster.x-k8s.io/deployment-name": string("md-0"), "topology.cluster.x-k8s.io/owned": string("")}},
                        ... // 2 identical entries
                },
                "status": map[string]any{"availableReplicas": int64(0), "conditions": []any{map[string]any{"lastTransitionTime": string("2022-07-28T14:02:57Z"), "message": string("Minimum availability requires 3 replicas, current 0 available"), "reason": string("WaitingForAvailableMachines"), "severity": string("Warning"), ...}, map[string]any{"lastTransitionTime": string("2022-07-28T14:02:57Z"), "message": string("Minimum availability requires 3 replicas, current 0 available"), "reason": string("WaitingForAvailableMachines"), "severity": string("Warning"), ...}}, "observedGeneration": int64(1), "phase": string("ScalingUp"), ...},
        },
  }
```

New output:
```
❯ cat ./out/modified/MachineDeployment_default_capi-quickstart-md-0-djlql.diff
--- out/modified/MachineDeployment_default_capi-quickstart-md-0-djlql.original.yaml     2022-07-27 18:52:24.000000000 +0200
+++ out/modified/MachineDeployment_default_capi-quickstart-md-0-djlql.modified.yaml     2022-07-27 18:52:24.000000000 +0200
@@ -73,13 +73,13 @@
    kind: Cluster
    name: capi-quickstart
    uid: 29e69ad2-e93f-45fe-bda0-06b0c7ed97b2
-  resourceVersion: "2116"
+  resourceVersion: "2"
  uid: aff0a885-dd95-4a64-8fee-c188b64bb4e4
spec:
  clusterName: capi-quickstart
  minReadySeconds: 0
  progressDeadlineSeconds: 600
-  replicas: 3
+  replicas: 2
  revisionHistoryLimit: 1
  selector:
    matchLabels:
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6985

### CODE REVIEW

File: cmd/clusterctl/cmd/topology_plan.go [hunk 1]
```diff
@@ -22,9 +22,9 @@ import (
 	"path"
 	"sort"
 
-	"github.com/google/go-cmp/cmp"
 	"github.com/olekukonko/tablewriter"
 	"github.com/pkg/errors"
+	"github.com/pmezard/go-difflib/difflib"
```
Comment (fabriziopandini) [inline cmd/clusterctl/cmd/topology_plan.go:27]: The project seems not maintained anymore.
Could you point me to how it is managed diff in kubectl so I can take a look?

Comment (chrischdi) [reply cmd/clusterctl/cmd/topology_plan.go:27]: kubectl makes use of an external `diff` binary via the `DiffProgram` struct:

https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/diff/diff.go#L213

Comment (chrischdi) [reply cmd/clusterctl/cmd/topology_plan.go:27]: Ready to take a look implementation: https://github.com/kubernetes-sigs/cluster-api/compare/main...chrischdi:cluster-api:pr-topology-plan-diff-binary?expand=1

Comment (fabriziopandini) [reply cmd/clusterctl/cmd/topology_plan.go:27]: I'm +1 to the proposed implementation that depends on diff command + allows override via an env variable
@vincepri opinions?

Comment (vincepri) [reply cmd/clusterctl/cmd/topology_plan.go:27]: +1 on the diff command dependency rather than depending on an unmaintained library

Comment (chrischdi) [reply cmd/clusterctl/cmd/topology_plan.go]: changed implementation to adapt the `kubectl` style :-)

**Comment by chrischdi** [comment]
changed implementation to adapt the `kubectl` style :-)

**Review by ykakarap — APPROVED** [review]
/lgtm

Works great!

**Comment by fabriziopandini** [comment]
Thanks @chrischdi for working on this issue!
/lgtm
/approve

**Pull request merged**